lua/nvim-treesitter/ts_utils.lua

```diff
@@ -212,4 +214,97 @@ function M.get_node_at_cursor(winnr)
   return root:named_descendant_for_range(cursor[1]-1,cursor[2],cursor[1]-1,cursor[2])
 end
 
+-- Finds the definition node and it's scope node of a node
+-- @param node starting node
+-- @param bufnr buffer
+-- @returns the definition node and the definition nodes scope node
+function M.find_definition(node, bufnr)
+  local bufnr = bufnr or api.nvim_get_current_buf()  
+  local node_text = M.get_node_text(node)[1]
+  local current_scope = M.containing_scope(node)
+  local _, _, node_start = node:start()
+
+  -- If a scope wasn't found then use the root node
+  if current_scope == node then
+    current_scope = parsers.get_parser(bufnr).tree:root()
+  end
+  
+  while current_scope ~= nil and current_scope ~= node do
+    for _, def in ipairs(locals.collect_locals(bufnr, current_scope)) do
+      if def.definition then
+        for _, def_node in ipairs(M.get_local_nodes(def.definition)) do
+          local _, _, def_start = def_node:start()
+
+          if M.get_node_text(def_node)[1] == node_text and def_start < node_start then
+            return def_node, current_scope
+          end
+        end
+      end
+    end
+
+    current_scope = M.containing_scope(current_scope:parent())
+  end
+
+  return nil, nil
+end
+
+-- Gets all nodes from a local list result.
+-- @param local_def the local list result
+-- @returns a list of nodes
+function M.get_local_nodes(local_def)
+  local result = {}
+
+  M.recurse_local_nodes(local_def, function(_, node) 
+    table.insert(result, node)
+  end)
+
+  return result
+end
+
+-- Recurse locals results until a node is found.
+-- The accumulator function is given
+-- * The table of the node
+-- * The node
+-- * The full definition match `@definition.var.something` -> 'var.something'
+-- * The last definition match `@definition.var.something` -> 'something'
+-- @param The locals result
+-- @param The accumulator function
+-- @param The full match path to append to
+-- @param The last match
+function M.recurse_local_nodes(local_def, accumulator, full_match, last_match)
+  if local_def.node then
+    accumulator(local_def, local_def.node, full_match, last_match)
+  else
+    for match_key, def in pairs(local_def) do
+      M.recurse_local_nodes(
+        def, 
+        accumulator, 
+        full_match and (full_match..'.'..match_key) or match_key, 
+        match_key)
+    end
+  end
+end
+
+-- Finds usages of a node in a particula scope
+-- @param node the node to find usages for
+-- @param scope_node the node to look within
+-- @returns a list of nodes
+function M.find_usages(node, scope_node)
+  local usages = {}
+  local node_text = M.get_node_text(node)[1]
+
+  if not node_text or #node_text < 1 then return {} end
+
+  for _, def in ipairs(locals.collect_locals(bufnr, scope_node)) do
```

Comment:
Not sure that you want to use `collect_locals` directly here, you might prefer `get_locals`, `collect_locals` was actually meant to be a hidden function.

Reply:
I can use the higher apis. Requires a little bit of refactoring but is probably for the better.

Reply:
But since you need to have that `scope_node` argument that will be better on the performances.